### PR TITLE
WINDUPRULE-109: Long Xpaths suffered from not executed parts

### PR DESCRIPTION
**Pull request by mbriskar**
This may be a rather complicated issue.
In the xpath, we generate multiple startFrame(index...) methods that register Maps in the cache. However, if the xpath is long, not all of the startFrame() are executed that means not all of the numbers are registered.

### CODE REVIEW

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-pr-builder/1385/

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://wonka.mw.lab.eng.bos.redhat.com/jenkins//job/windup-core-PR-dependents-check/757/

File: rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFile.java [hunk 1]
```diff
@@ -574,7 +574,7 @@ public Object evaluate(@SuppressWarnings("rawtypes") List args) throws XPathFunc
                     }
                 }
 
-                for (Map.Entry<String, String> entry : paramMatchCache.getVariables(frameIdx).entrySet())
+                for (Map.Entry<String, String> entry : paramMatchCache.getVariables().entrySet())
```
Comment (mbriskar) [inline rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFile.java]: I didn't get the reason why frameIdx was put there. Maybe I missed something

File: rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFileParameterMatchCache.java [hunk 2]
```diff
@@ -37,10 +41,10 @@ public void addVariable(int frameID, String key, String value)
         vars.get(frameID).put(key, value);
     }
 
-    public Map<String, String> getVariables(int frameID)
+    public Map<String, String> getVariables()
```
Comment (mbriskar) [inline rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFileParameterMatchCache.java]: or maybe I should have included only up to frameID?

File: rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFileNamespace.java [hunk 3]
```diff
@@ -16,4 +16,12 @@
...
      */
     ConditionBuilder as(String variable);
+
+    /**
+     * Specify the namespace used in the xpath that was provided sooner. Use this now only if you don't want to specify any of these: xpath result regex, DTD regex, file naem regex.
+     * @param prefix the namespace prefix
+     * @param url the namespace url
+     * @return the last step in XmlFile definition to provide the variable name
+     */
+    XmlFileNamespace namespace(String prefix, String url);
```
Comment (mbriskar) [inline rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFileNamespace.java]: added this so we can add more namespaces for XmlFile.namespace().namespace()...

File: rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFile.java (hunk 1 above)
Comment (jsight) [inline rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFile.java]: In looking at the code, I'm not sure either. I think removing it is probably the right thing to do.

File: rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFileNamespace.java (hunk 3 above)
Comment (jsight) [inline rules-xml/api/src/main/java/org/jboss/windup/rules/apps/xml/condition/XmlFileNamespace.java]: Good. :)

**Comment by jsight** [comment]
Loooks good to me... nice work and the test is nice!

**Pull request merged**